kubelet: add RawPodStatus and related functions

yujuhong opened the pull request:
RawPodStatus will be the internal status of the pod that kubelet relies on for syncing
This PR only adds the type and functions. The followup PRs will start implementing them.

This is part of #12619.

yujuhong committed:
kubelet: add RawPodStatus and related functions

RawPodStatus will be the internal status of the pod that kubelet relies on for
syncing.

yujuhong commented:
Note that the first commit is part of #13571. Please review only the second commit.

/cc @dchen1107 @yifan-gu @Random-Liu

yifan-gu commented:
LGTM， except small nits

Random-Liu commented:
LGTM.
It seems that Jenkins is broken again...
@k8s-bot test this please.

yujuhong commented:
Addressed comments. Also added comments to remove ConvertRawToPodStatus in the future after discussing with @yifan-gu offline. Still waiting for #13571 to land though.

yifan-gu commented:
Rebase on this as #13571 is merged?

yujuhong commented:
Rebased and squashed the commits.

yujuhong commented:
@k8s-bot test this

yujuhong commented:
Change to use pointers for consistency. Applying lgtm based on @yifan-gu's LGTM

yujuhong commented:
Actually, @dchen1107, do you want to take a look?

dchen1107 commented on pkg/kubelet/container/runtime.go:
```diff
@@ -88,9 +89,19 @@ type Runtime interface {
-	// statuses in a pod always have a deterministic ordering (eg: sorted by name).
+	// all containers in the pod. Clients of this interface assume the
+	// containers' statuses in a pod always have a deterministic ordering
+	// (e.g., sorted by name).
```
Please add comment that this is going to be deprecated, and instead the GetRawPodStatus is the one to use for all runtime implementations.

dchen1107 commented on pkg/kubelet/container/runtime.go:
```diff
@@ -88,9 +89,19 @@ type Runtime interface {
 	GetPodStatus(*api.Pod) (*api.PodStatus, error)
+	// GetRawPodStatus retrieves the status of the pod, including the
+	// information of all containers in the pod that are visble in Runtime.
+	GetRawPodStatus(uid types.UID, name, namespace string) (*RawPodStatus, error)
```
Please comment this method is going to be rename to GetPodStatus along with deprecating previous GetPodStatus defined above.

dchen1107 commented on pkg/kubelet/container/runtime.go:
```diff
@@ -88,9 +89,19 @@ type Runtime interface {
+	// pod-level information for api.PodStatus (e.g., check whether the image
+	// exists to determine the reason). We should try generalizing the logic
+	// for all container runtimes in kubelet and remove this funciton.
+	ConvertRawToPodStatus(*api.Pod, *RawPodStatus) (*api.PodStatus, error)
```
Comment this one is to-be-deprecated.

dchen1107 commented:
LGTM 

cc/ @feiskyer

yujuhong commented:
@dchen1107, I added all the TODOs as requested. I modified the original commit to avoid yet another rebase since this is a relatively small change. Please apply the lgtm label if you ok with this PR. Thanks!

/cc @Random-Liu

feiskyer commented:
LGTM

dchen1107 commented:
LGTM

yifan-gu commented:
> GCE e2e build/test failed for commit 760309f.

I saw similar e2e failues in my PR https://github.com/kubernetes/kubernetes/pull/17240#issuecomment-157572665

@yujuhong Do you have ideas what's going on here? Seems the cluster is not reachable suddenly for some reason.

yujuhong commented:
@yifan-gu, I think it's an infra issue and we have tests leaking GCE resources too.

yujuhong commented:
@k8s-bot test this

yujuhong commented:
@k8s-bot test this

yujuhong commented:
@k8s-bot test this

yujuhong commented:
@dchen1107, this PR doesn't touch existing code. Can we manually merge it?
